kvserver: count draining nodes as live when computing quorum

aayushshah15 opened the pull request:
Similar to #67714

Draining nodes were considered non-live by the allocator when it made the
determination of whether a range could achieve quorum. This meant that, for
instance, on a cluster with a replication factor of 5, if we had 3 or more
nodes marked draining, we (with a high likelihood) wouldn't be able to
decommission _any other_ nodes from the cluster.

Furthermore, due to the same reason as above the system also would incorrectly
decide to not rebalance ranges that had more than a quorum of replicas on
draining nodes.

This patch fixes this problem by considering replicas on draining nodes as live
for the purposes of determining whether a range has quorum. This likely fixes a
subset of "stuck decommissioning" issues we've seen in the wild.

Follows from cockroachlabs/support#1105

Release justification: bug fix

Release note(bug fix): Previously, draining a quorum of nodes (i.e. >=2 if
replication factor is 3, >=3 if replication factor is 5, etc) would block the
subsequent decommissioning of any other nodes in the cluster. This patch fixes
this bug. Now, if the lowest replication factor of some zone in the cluster is
RF, operators should be able to safely drain up to RF-1 nodes simultaneously.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68652)
<!-- Reviewable:end -->

aayushshah15 committed:
kvserver: count draining nodes as live when computing quorum

Similar to https://github.com/cockroachdb/cockroach/pull/67714

Draining nodes were considered non-live by the allocator when it made the
determination of whether a range could achieve quorum. This meant that, for
instance, on a cluster with a replication factor of 5, if we had 3 or more
nodes marked draining, we (with a high likelihood) wouldn't be able to
decommission nodes from the cluster.

Furthermore, due to the same reason as above the system also would incorrectly
decide to not rebalance ranges that had more than a quorum of replicas on
draining nodes.

This patch fixes this problem by considering replicas on draining nodes as live
for the purposes of determining whether a range has quorum. This likely fixes a
considerable subset of "stuck decommissioning" issues we've seen in the wild.

Follows from https://github.com/cockroachlabs/support/issues/1105

Release note: None

aayushshah15 committed:
roachtest: add a drain+decommission roachtest

This commit adds a roachtest that's meant to be a regression test against the
hazard addressed by the first commit in this PR.

This roachtest is meant to ensure that nodes that are marked "draining" are
considered "live" by the allocator for when it makes the determination of
whether a range can achieve quorum.

Release note: None

aayushshah15 committed:
kvserver: stop transferring leases to draining/suspect nodes in the StoreRebalancer

This commit prevents the StoreRebalancer from transferring leases to replicas
on draining or suspect nodes. In some cases, we've seen this to cause new
leases to be pushed to nodes that take too long to drain or that are stuck
while draining due to other bugs.

Informs https://github.com/cockroachlabs/support/issues/1105

Release note: None

erikgrinaker reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo test failures

> Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-:-MgytfQnAVBk6brEzbDy:b-cm6ptw)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
This mostly <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, though I'm a little unclear on why we want draining nodes to be considered live for the purposes of computing quorum. Is my understanding correct that once a node has finished draining, it will terminate and become non-live? If so, then why are we finding ourselves in states where we have 3 nodes that are draining and also have a node performing a decommission and we need these draining nodes to contribute to quorum? If the nodes completed draining and terminated then they certainly wouldn't contribute to quorum, so aren't the expectations here racy? Put differently, if a draining node can at any moment finish draining and terminate, then isn't a draining node as good as dead?

> Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-:-MgzxkHt3t7axOSwGjlQ:b-sey176)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[pkg/kv/kvserver/store_pool.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-Mgzwch652XDh_gw64bR:-Mgzwch652XDh_gw64bS:bbl8p0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbc4f4d1118abbe8056f20584126a36c7c2a75/pkg/kv/kvserver/store_pool.go#L706)):*
> ```Go
> 		// No-op.
> 		case storeStatusSuspect, storeStatusDraining:
> 			if includeSuspectAndDrainingStores {
> ```

Did any unit tests pick up this change? If not, are we missing some test coverage?
___
*[pkg/kv/kvserver/store_rebalancer.go, line 429 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-Mgzw19q8_ETLXHbPCwF:-Mgzw19q8_ETLXHbPCwG:b-gns1pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f83ebcf483fd5bc4ac8f785e087783bb9ffc49e/pkg/kv/kvserver/store_rebalancer.go#L429)):*
> ```Go
> 		preferred := sr.rq.allocator.preferredLeaseholders(zone, candidates)
> 
> 		// Filter both the list of preferred stores as well as the list of all
> ```

Do we have a way to unit test this change in behavior? Ideally in something targetted like `TestChooseLeaseToTransfer`.
___
*[pkg/kv/kvserver/store_rebalancer.go, line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MgzvmMD78aabaNpT8Rx:-MgzvmMEBSWANUKnw1jR:brubu80) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f83ebcf483fd5bc4ac8f785e087783bb9ffc49e/pkg/kv/kvserver/store_rebalancer.go#L434)):*
> ```Go
> 		const includeSuspectAndDrainingStores = false
> 		preferred, _ = sr.rq.allocator.storePool.liveAndDeadReplicas(preferred, includeSuspectAndDrainingStores)
> 		candidates, _ = sr.rq.allocator.storePool.liveAndDeadReplicas(candidates, includeSuspectAndDrainingStores)
> ```

Were we previously able to transfer the lease to dead nodes as well? I don't see what was preventing that.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
> Is my understanding correct that once a node has finished draining, it will terminate and become non-live? 

The nodes wont terminate once the draining process completes, they stay in this state of not accepting new leases/replicas and new SQL connections until they're either restarted or terminated explicitly. 

One of our customers had a workflow where if they had to remove, say 10 nodes from the cluster, they would _first_ mark all these 10 nodes as draining and then gradually decommission them (this was to avoid moving replicas to node that are soon to be decommissioned anyway). One of the main reasons this decommissioning would get stuck is because the allocator would incorrectly think that a bunch of ranges were unavailable (and thus would not take any action on them).

You're right about the lacking unit test coverage, I'll update the PR with some.

nvanbenschoten commented:
Got it, thanks for the explanation. In that case, I agree with the changes that this PR is making.

knz commented:
I really like this PR a lot.
I also think it needs a release note, because it answers a question that @taroface had a few months ago: *"what would be a good use case for a separate drain command?"*

The explanation you provide above is such a good use case, and it deserves to be in the release note so the docs team can pick it up once this change is in.

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @erikgrinaker, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_command.go, line 2047 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEj-m773vuC5sfpVGy:-MhEj-m773vuC5sfpVGz:b-ckzxv7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/replica_command.go#L2047)):*
> Otherwise, they are excluded from the returned slices.

Is it true that replicas can be excluded from both results? So, if this flag is not passed, the caller needs to consider three categories of replicas - live, dead, neither, with the latter not being explicitly represented? This seems like a weird contract to me. Would any caller complain if the suspect/draining replicas were included in the `deadReplicas` slice?
___
*[pkg/kv/kvserver/store_pool.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEjyaEAc5ZZ4OfX_5c:-MhEjyaEAc5ZZ4OfX_5d:b-7gp7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L239)):*
> ```Go
> 	storeStatusAvailable
> 	// The store is decommissioning.
> 	storeStatusDecommissioning
> ```

maybe take the opportunity to explain the priority between this and dead/suspect/draining.
___
*[pkg/kv/kvserver/store_pool.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEjqzX-UDfAxAqwiud:-MhEjqzX-UDfAxAqwiue:b-qxeekq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L240)):*
> ```Go
> 	// The store is decommissioning.
> 	storeStatusDecommissioning
> 	// The store failed it's liveness heartbeat recently and is considered
> ```

maybe take the opportunity to explain that stores move from dead to suspicious.
___
*[pkg/kv/kvserver/store_pool.go, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhElYj63Nlx7Y6jGTc-:-MhElYj63Nlx7Y6jGTc0:b-demr21) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L833)):*
> ```Go
> // the required attributes and their associated stats. The storeList is filtered
> // according to the provided storeFilter. It also returns the total number of
> // alive and throttled stores.
> ```

The comment on returning the number of throttled stores seems obsolete. Perhaps take the opportunity to explain the last retval; it seems like a weird return value because it doesn't include store ids.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
>The explanation you provide above is such a good use case

Even with this patch, we cannot actually support draining >=  RF nodes at the same time where RF is the lowest replication factor of some zone in the cluster. This is because _decommissioning_ any of these draining nodes will require lease transfers away from the decommissioning node (since the leaseholders on the decommissioning node cannot remove themselves). These lease transfers need to succeed, so every range needs to have at least one replica on a non-draining node. 

This patch does let the user simultaneously drain RF-1 nodes from their cluster though. I've amended the release note to reflect this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-:-Mha5Qmk5OLdASc9NzLM:bt4zwoq)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei, @erikgrinaker, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_command.go, line 2047 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEj-m773vuC5sfpVGy:-MhP6XTx3lZDd7SCSL3T:b-lip28m) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/replica_command.go#L2047)):*
>This seems like a weird contract to me. Would any caller complain if the suspect/draining replicas were included in the deadReplicas slice?

The callers use that `deadReplicas` slice essentially as the list of replicas that need to be replaced -- which is not what we'd like to do for replicas on draining or suspect nodes. So in this case, it feels like we do need the disambiguation.
___
*[pkg/kv/kvserver/store_pool.go, line 706 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-Mgzwch652XDh_gw64bR:-Mha4LN55oETEXswSW42:b-90p8me) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfbc4f4d1118abbe8056f20584126a36c7c2a75/pkg/kv/kvserver/store_pool.go#L706)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Did any unit tests pick up this change? If not, are we missing some test coverage?
</blockquote></details>

Added a short new test for it. 
___
*[pkg/kv/kvserver/store_pool.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEjyaEAc5ZZ4OfX_5c:-Mha44yWCV4_Md30B0HP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L239)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe take the opportunity to explain the priority between this and dead/suspect/draining.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_pool.go, line 240 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEjqzX-UDfAxAqwiud:-Mha4JaDBi8cxJUwQXjx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L240)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe take the opportunity to explain that stores move from dead to suspicious.
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/store_pool.go, line 833 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhElYj63Nlx7Y6jGTc-:-Mha54dd3qHaSpXgdjXt:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/store_pool.go#L833)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The comment on returning the number of throttled stores seems obsolete. Perhaps take the opportunity to explain the last retval; it seems like a weird return value because it doesn't include store ids.
</blockquote></details>

Fixed.
___
*[pkg/kv/kvserver/store_rebalancer.go, line 429 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-Mgzw19q8_ETLXHbPCwF:-Mha5DXEATWFtuOiX8jZ:b-6b6l7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f83ebcf483fd5bc4ac8f785e087783bb9ffc49e/pkg/kv/kvserver/store_rebalancer.go#L429)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we have a way to unit test this change in behavior? Ideally in something targetted like `TestChooseLeaseToTransfer`.
</blockquote></details>

`TestChooseLeaseToTransfer` is getting rehauled in https://github.com/cockroachdb/cockroach/pull/65379. Would you mind if I handled it there?
___
*[pkg/kv/kvserver/store_rebalancer.go, line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MgzvmMD78aabaNpT8Rx:-Mha56XR5exU4i_myhc0:b-4avj6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f83ebcf483fd5bc4ac8f785e087783bb9ffc49e/pkg/kv/kvserver/store_rebalancer.go#L434)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Were we previously able to transfer the lease to dead nodes as well? I don't see what was preventing that.
</blockquote></details>

Yeah, nothing was preventing that. 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> One of our customers had a workflow where if they had to remove, say 10 nodes from the cluster, they would first mark all these 10 nodes as draining and then gradually decommission them (this was to avoid moving replicas to node that are soon to be decommissioned anyway). One of the main reasons this decommissioning would get stuck is because the allocator would incorrectly think that a bunch of ranges were unavailable (and thus would not take any action on them).

Please include this paragraph in the commit message somehow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-:-MhtmhzeDCAbhfeMKD0h:b-ptqc7n)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @aayushshah15, @erikgrinaker, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_command.go, line 2047 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEj-m773vuC5sfpVGy:-MhthSDR28nDjk1WbaVr:b-9e40gi) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/replica_command.go#L2047)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

>This seems like a weird contract to me. Would any caller complain if the suspect/draining replicas were included in the deadReplicas slice?

The callers use that `deadReplicas` slice essentially as the list of replicas that need to be replaced -- which is not what we'd like to do for replicas on draining or suspect nodes. So in this case, it feels like we do need the disambiguation.
</blockquote></details>

Ack on the distinction between dead and suspect/draining.
I won't make a big deal out of it, but I'd still encourage figuring out a different format for the returned values. Perhaps get rid of `includeSuspectStores` and return a `replicasWithStatus` object with filtering methods that let the caller more explicitly choose what they're looking at.
___
*[pkg/kv/kvserver/store_pool.go, line 239 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhtiHhr0FCVpKV50GtC:-MhtiHhr0FCVpKV50GtD:bodd73c) ([raw file](https://github.com/cockroachdb/cockroach/blob/3235374a545748af2ce9ad7aae85d82fe3abcf2c/pkg/kv/kvserver/store_pool.go#L239)):*
> ```Go
> 	storeStatusAvailable
> 	// The store is decommissioning. If draining or suspect stores are
> 	// decommissioned, this status takes precedence over `storeStatusDraining`
> ```

nit: s/decommissioned/decommissioning


<!-- Sent from Reviewable.io -->

andreimatei commented:
how come the 2nd commit doesn't close cockroachlabs/support#1105 ?

aayushshah15 reviewed (commented):
> how come the 2nd commit doesn't close cockroachlabs/support#1105 ?

I think its possible we still have an undiagnosed issue with draining nodes that are failing liveness heartbeats. I think https://github.com/cockroachdb/cockroach/pull/67714 plus this patch may solve it but I'd like to test it out with that customer before we close that ticket.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-:-MiMbtaI2N8L_aaWLI5q:b-s73jm4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @andreimatei, @erikgrinaker, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_command.go, line 2047 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhEj-m773vuC5sfpVGy:-MiMesYgC7YG59NFMXyA:b-93t5c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5647775aa86b66cf0f9927d4b92bca9949799daa/pkg/kv/kvserver/replica_command.go#L2047)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Ack on the distinction between dead and suspect/draining.
I won't make a big deal out of it, but I'd still encourage figuring out a different format for the returned values. Perhaps get rid of `includeSuspectStores` and return a `replicasWithStatus` object with filtering methods that let the caller more explicitly choose what they're looking at.
</blockquote></details>

:+1: Agreed. Added a TODO for now since there's a ton of callers for this function and I'd like to keep the patch small.
___
*[pkg/kv/kvserver/store_pool.go, line 239 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/68652#-MhtiHhr0FCVpKV50GtC:-MiMbh_R6PQB-3UYYIbh:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3235374a545748af2ce9ad7aae85d82fe3abcf2c/pkg/kv/kvserver/store_pool.go#L239)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/decommissioned/decommissioning
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->

aayushshah15 commented:
CI was failing due to missing release justification.

TFTRs!

bors r+
